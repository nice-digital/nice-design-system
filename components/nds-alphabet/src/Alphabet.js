import React, { Children, isValidElement, cloneElement } from "react";
import PropTypes from "prop-types";
import "../scss/alphabet.scss";

export const Alphabet = props => {
	const { children, id, chunky, ...attrs } = props;

	const chunkyChild = child => {
		return isValidElement(child)
			? cloneElement(child, { chunky: true })
			: child;
	};

	return (
		<ol
			className={`alphabet ${chunky ? "alphabet--chunky" : ""}`}
			id={id || "a-to-z"}
			{...attrs}
		>
			{chunky ? Children.map(children, child => chunkyChild(child)) : children}
		</ol>
	);
};

Alphabet.propTypes = {
	id: PropTypes.string,
	children: PropTypes.node.isRequired,
	chunky: PropTypes.bool
};

export const Letter = props => {
	const { children, label, to, chunky, elementType, ...attrs } = props;

	let body;

	if (!to) {
		// No link
		body = <span aria-label={label}>{children}</span>;
	} else if (to[0] === "#") {
		// Link to an id on the same page
		body = (
			<a href={to} aria-label={label}>
				{children}
			</a>
		);
	} else {
		// Link to somewhere else!
		body = (
			<Link
				elementType={elementType}
				ariaLabel={label}
				destination={to}
				text={children}
			/>
		);
	}

	return (
		<li
			className={`alphabet__letter ${chunky ? "alphabet__letter--chunky" : ""}`}
			{...attrs}
		>
			{body}
		</li>
	);
};

Letter.propTypes = {
	children: PropTypes.node.isRequired,
	chunky: PropTypes.bool,
	label: PropTypes.string,
	to: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
	elementType: PropTypes.elementType
};

const Link = ({
	text,
	destination,
	elementType: ElementType = "a",
	method,
	ariaLabel
}) => {
	let linkProps = {
		className: "alphabet__link",
		[method || (ElementType === "a" && "href") || "to"]: destination
	};

	return (
		<span className="alphabet__link-wrapper">
			<ElementType aria-label={ariaLabel} {...linkProps}>
				{text}
			</ElementType>{" "}
		</span>
	);
};

Link.propTypes = {
	text: PropTypes.string,
	destination: PropTypes.string,
	elementType: PropTypes.elementType,
	method: PropTypes.string,
	ariaLabel: PropTypes.string
};
