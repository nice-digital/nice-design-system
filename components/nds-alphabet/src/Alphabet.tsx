import React, { Children, isValidElement, cloneElement } from "react";
import "../scss/alphabet.scss";

export interface AlphabetProps {
	id?: string;
	children: React.ReactElement<LetterProps>[];
	chunky?: boolean;
	className?: string;
	[key: string]: unknown;
}

export interface LetterProps {
	children: React.ReactNode;
	label?: string;
	chunky?: boolean;
	elementType?: React.ElementType;
	to?: string | false;
	[key: string]: unknown;
}

export interface LinkProps {
	text: React.ReactNode;
	destination: string;
	elementType?: React.ElementType;
	method?: string;
	ariaLabel?: string;
}

export const Alphabet: React.FC<AlphabetProps> = (props: AlphabetProps) => {
	const { children, className, id, chunky, ...attrs } = props;

	const chunkyChild = (child: React.ReactElement<LetterProps>) => {
		return isValidElement(child)
			? cloneElement(child, { chunky: true })
			: child;
	};

	return (
		<ol
			className={[
				`alphabet ${chunky ? "alphabet--chunky" : ""}`,
				className
			].join(" ")}
			id={id || "a-to-z"}
			{...attrs}
		>
			{chunky
				? Children.map(children, (child) => chunkyChild(child))
				: children}
		</ol>
	);
};

export const Letter: React.FC<LetterProps> = (props: LetterProps) => {
	const { children, label, to, chunky, elementType, ...attrs } = props;

	let body;

	if (!to) {
		// No link
		body = <span aria-label={label}>{children}</span>;
	} else if (to[0] === "#") {
		// Link to an id on the same page
		body = (
			<a href={to} aria-label={label}>
				{children}
			</a>
		);
	} else {
		// Link to somewhere else!
		body = (
			<Link
				elementType={elementType}
				ariaLabel={label}
				destination={to}
				text={children}
			/>
		);
	}

	return (
		<li
			className={`alphabet__letter ${chunky ? "alphabet__letter--chunky" : ""}`}
			{...attrs}
		>
			{body}
		</li>
	);
};

const Link: React.FC<LinkProps> = ({
	text,
	destination,
	elementType: ElementType = "a",
	method,
	ariaLabel
}: LinkProps) => {
	let linkProps = {
		className: "alphabet__link",
		[method || (ElementType === "a" && "href") || "to"]: destination
	};

	return (
		<ElementType aria-label={ariaLabel} {...linkProps}>
			{text}
		</ElementType>
	);
};
