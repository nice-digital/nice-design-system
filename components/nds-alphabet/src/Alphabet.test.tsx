import React from "react";
import { render } from "@testing-library/react";
import { Alphabet, Letter } from "./Alphabet";

describe("Alphabet", () => {
	const allLetters = "abcdefghijklmnopqrstuvwxyz".split("");
	const letterElements = allLetters.map((letter) => (
		<Letter key={letter} to={`#${letter}`} label={`Letter ${letter}`}>
			{letter.toUpperCase()}
		</Letter>
	));

	it("should match snapshot", () => {
		const wrapper = render(<Alphabet>{letterElements}</Alphabet>);
		expect(wrapper).toMatchSnapshot();
	});

	it("should add a class of 'alphabet' to the default variant", () => {
		const wrapper = render(<Alphabet>{letterElements}</Alphabet>);
		expect(wrapper.getByRole("list")).toHaveClass("alphabet");
	});

	it("should add a custom class when one is specified", () => {
		const wrapper = render(
			<Alphabet className="test-class">{letterElements}</Alphabet>
		);
		expect(wrapper.getByRole("list")).toHaveClass("test-class");
	});

	it("should add a class of 'alphabet--chunky' to the chunky variant", () => {
		const wrapper = render(<Alphabet chunky>{letterElements}</Alphabet>);
		expect(wrapper.getByRole("list")).toHaveClass("alphabet--chunky");
	});

	it("should render a child component for each letter of the alphabet", () => {
		const wrapper = render(<Alphabet>{letterElements}</Alphabet>);
		expect(wrapper.getByRole("list").children.length).toBe(allLetters.length);
	});

	it("should render chunky child components when the chunky variant is specified on the parent", () => {
		const wrapper = render(<Alphabet chunky>{letterElements}</Alphabet>);
		expect(wrapper.getAllByRole("listitem")[0]).toHaveClass(
			"alphabet__letter alphabet__letter--chunky"
		);
	});

	it("should append className prop to rendered class attribute", () => {
		const wrapper = render(
			<Alphabet className="test">{letterElements}</Alphabet>
		);
		expect(wrapper.getByRole("list")).toHaveClass("test");
	});

	it("should render additional props as attributes", () => {
		const wrapper = render(
			<Alphabet aria-label="hello">{letterElements}</Alphabet>
		);
		expect(wrapper.getByRole("list").getAttribute("aria-label")).toBe("hello");
	});
});
