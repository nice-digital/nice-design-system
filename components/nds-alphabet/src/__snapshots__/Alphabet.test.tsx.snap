// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alphabet should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ol
        class="alphabet  "
        id="a-to-z"
      >
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter a"
            href="#a"
          >
            A
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter b"
            href="#b"
          >
            B
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter c"
            href="#c"
          >
            C
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter d"
            href="#d"
          >
            D
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter e"
            href="#e"
          >
            E
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter f"
            href="#f"
          >
            F
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter g"
            href="#g"
          >
            G
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter h"
            href="#h"
          >
            H
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter i"
            href="#i"
          >
            I
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter j"
            href="#j"
          >
            J
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter k"
            href="#k"
          >
            K
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter l"
            href="#l"
          >
            L
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter m"
            href="#m"
          >
            M
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter n"
            href="#n"
          >
            N
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter o"
            href="#o"
          >
            O
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter p"
            href="#p"
          >
            P
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter q"
            href="#q"
          >
            Q
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter r"
            href="#r"
          >
            R
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter s"
            href="#s"
          >
            S
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter t"
            href="#t"
          >
            T
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter u"
            href="#u"
          >
            U
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter v"
            href="#v"
          >
            V
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter w"
            href="#w"
          >
            W
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter x"
            href="#x"
          >
            X
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter y"
            href="#y"
          >
            Y
          </a>
        </li>
        <li
          class="alphabet__letter "
        >
          <a
            aria-label="Letter z"
            href="#z"
          >
            Z
          </a>
        </li>
      </ol>
    </div>
  </body>,
  "container": <div>
    <ol
      class="alphabet  "
      id="a-to-z"
    >
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter a"
          href="#a"
        >
          A
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter b"
          href="#b"
        >
          B
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter c"
          href="#c"
        >
          C
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter d"
          href="#d"
        >
          D
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter e"
          href="#e"
        >
          E
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter f"
          href="#f"
        >
          F
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter g"
          href="#g"
        >
          G
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter h"
          href="#h"
        >
          H
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter i"
          href="#i"
        >
          I
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter j"
          href="#j"
        >
          J
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter k"
          href="#k"
        >
          K
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter l"
          href="#l"
        >
          L
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter m"
          href="#m"
        >
          M
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter n"
          href="#n"
        >
          N
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter o"
          href="#o"
        >
          O
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter p"
          href="#p"
        >
          P
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter q"
          href="#q"
        >
          Q
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter r"
          href="#r"
        >
          R
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter s"
          href="#s"
        >
          S
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter t"
          href="#t"
        >
          T
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter u"
          href="#u"
        >
          U
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter v"
          href="#v"
        >
          V
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter w"
          href="#w"
        >
          W
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter x"
          href="#x"
        >
          X
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter y"
          href="#y"
        >
          Y
        </a>
      </li>
      <li
        class="alphabet__letter "
      >
        <a
          aria-label="Letter z"
          href="#z"
        >
          Z
        </a>
      </li>
    </ol>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
